doc/adr004-effect-system.md

```diff
@@ -38,150 +36,292 @@ simplicity.
+
 ```
 
-Other forms are actually sugaring for these two. Effects like `Update['x'] &
-Update['y']` are in an invalid format and should be simplified before applying
-any other transformation. See equivalence rules below.
+The effect `Read['x', 'y'] & Update[v]` would be represented as:
 
-The motivation for this form is to help writing effect signatures for operators
-that care only about the read or the update part of some effect. For example,
-the `or` operator representing disjunction in Quint takes expressions with the
-identical updates, but doesn't have any restrictions on the read part. Ensuring
-this normal form allows us to write its signature as `(Read[r1] & Update[u],
-Read[r2] & Update[u]) => Read[r1 ∪ r2] & Update[u]`.
+```typescript
+{
+  kind: 'concrete',
+  components: [
+    { kind: 'read', entity: { kind: 'concrete', stateVariables: [{ name: 'x', reference: 1n }, { name: 'y', reference: 2n }] } },
+    { kind: 'update', entity: { kind: 'variable', name: 'v' } },
+  ],
+}
+```
 
-### Equivalence rules
+Effect signatures for operators may depend on its parameters. For that purpose,
+we also define arrow effects, which carry a list of effects for the respective
+parameters of an operator and the resulting effect.
 
-These are some equivalence rules to be used alongside unification, but that
-don't require any substitution. These are applied in a simplification process
-with the goal of reaching the normal form. Equivalence between `E1` and `E2` is
-expressed by `E1 ≡ E2`, and the equivalence symbol `≡` has the lowest precedence
-on this system.
+```typescript
+/* Arrow effects for expressions with effects depending on parameters */
+export interface ArrowEffect { kind: 'arrow', params: Effect[], result: Effect }
+```
 
+Just like with entities, an effect can also be a variable, which can be
+substituted with another effect during inferrence. Those are the most general
+kinds of effects, and will be used when no effect information for a given
+expression is available, i.e. `def foo(x) = x` will have a `(e) => e` effect
+signature, where `e` is an effect variable, since there are no constraints on
+the effect of `x`. An effect variable is equivalent to `Read[v1] & Update[v2] &
+Temporal[v3]`, where `v1`, `v2` and `v3` are entity variables.
+
+```typescript
+/* A variable representing some effect */
+export interface EffectVariable { kind: 'variable', name: string }
 ```
-E1 ≡ Ex & Read[x0, ..., xn]   E2 ≡ Ey & Read[y0, ..., yn]
----------------------------------------------------------- (READ)
-E1 & E2 ≡ Ex & Ey & Read[x0, ..., xn, y0, ..., yn]
 
+### Effect Schemes
+
+Both effects and entities can be universally quantified, and we need to keep
+track of which of them are. Effect schemes are responsible to keep track of that
+information.
+
+```typescript
+/*
+ * An effect scheme, listing which effect variables (names referring to effects)
+ * and entity variables (names refering to entities) are universally quantified
```

Comment:
I think this also confuses levels a bit. IIUC, the names here refer to variables, which stand for unknown effects, they do not refer to effects directly.

```suggestion
 * An effect scheme, listing which effect variables (represented by the names of effect variables)
 * and entity variables (represented by the names of entity variables) are universally quantified
```